src/components/input-number/input-number.tsx

```diff
@@ -0,0 +1,135 @@
+import React, { useMemo } from 'react'
+import clsx from 'clsx'
+import { ChevronDownOutline, ChevronUpOutline } from 'components/icons'
+import { useSyncedState } from 'hooks/useSyncedState'
+
+/**
+ * InputNumber properties
+ */
+export type InputNumberProps = Omit<
+  React.DetailedHTMLProps<
+    React.InputHTMLAttributes<HTMLInputElement>,
+    HTMLInputElement
+  >,
+  'step' | 'className'
+> & {
+  /** The number to which the current value is increased or decreased. It can be an integer or decimal */
+  step?: number
+  value?: number
+  defaultValue?: number
+  min?: number
+  max?: number
+  precision?: number
+  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void
+  /** Additional class applied to the input element */
+  className?: string
+  style?: React.CSSProperties
+}
+
+/** Enter a number within certain range with the mouse or keyboard. */
+export function InputNumber({
+  step = 1,
+  value,
+  defaultValue,
+  min,
+  max,
+  precision,
+  onChange,
+  className,
+  ...restProps
+}: InputNumberProps) {
+  const [inputValue, setInputValue] = useSyncedState(
+    (value || defaultValue) ?? 0,
+  )
+
+  const getPrecision = useMemo(() => {
+    if (precision) {
+      return precision
+    }
+    const stepString = String(step)
+    let stepPrecision = 0
+    if (stepString.indexOf('.') >= 0) {
+      stepPrecision = stepString.length - stepString.indexOf('.') - 1
```

Comment:
Wrap this logic into a helper function `getPrecisionFromNumber` and add comments for the logic over there.